HW2/P2/parallel_vector.pyx

```diff
@@ -54,39 +54,67 @@ cpdef move_data_serial(np.int32_t[:] counts,
                        np.int32_t[:] src,
                        np.int32_t[:] dest,
                        int repeat):
-   cdef:
-       int idx, r
+    cdef:
+        int idx, r
+
+    assert src.size == dest.size, "Sizes of src and dest arrays must match"
+    with nogil:
+        for r in range(repeat):
+            for idx in range(src.shape[0]):
+                if counts[src[idx]] > 0:
+                    counts[dest[idx]] += 1
+                    counts[src[idx]] -= 1
+
 
-   assert src.size == dest.size, "Sizes of src and dest arrays must match"
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+cdef void lock_and_increment(int idx, omp_lock_t *lock, int increment_by, np.int32_t[:] counts) nogil:
+    acquire(lock)
+    counts[idx] += increment_by
+    release(lock)
 
 
 cpdef move_data_fine_grained(np.int32_t[:] counts,
                              np.int32_t[:] src,
                              np.int32_t[:] dest,
                              int repeat):
-   cdef:
-       int idx, r
-       omp_lock_t *locks = get_N_locks(counts.shape[0])
+    cdef:
+        int idx, r, src_idx, dest_idx
+        omp_lock_t *locks = get_N_locks(counts.shape[0])
+        omp_lock_t *min_lock
+        omp_lock_t *max_lock
+
+    ##########s
+    # Your code here
+    # Use parallel.prange() and a lock for each element of counts to parallelize
+    # data movement.  Be sure to avoid deadlock, and double-locking.
+    ##########
+    with nogil:
+        for r in range(repeat):
+            for idx in prange(src.shape[0], num_threads=4):
+
+                # checking the locks instead of ids allows this code to
+                # work almost unchanged for coarse-grained locking
+                min_lock = &locks[src[idx]] if src[idx] < dest[idx] else &locks[dest[idx]]
+                max_lock = &locks[src[idx]] if src[idx] > dest[idx]  else &locks[dest[idx]]
 
-   ##########
-   # Your code here
-   # Use parallel.prange() and a lock for each element of counts to parallelize
-   # data movement.  Be sure to avoid deadlock, and double-locking.
-   ##########
-   with nogil:
-       for r in range(repeat):
-           for idx in range(src.shape[0]):
-               if counts[src[idx]] > 0:
-                   counts[dest[idx]] += 1
-                   counts[src[idx]] -= 1
+                # acquire the min_lock first
+                acquire(min_lock)
```

Comment:
what if min_lock or max_lock changes between the time of the condition checking and the time that you acquire the lock?